samples: adding imu sensor fusion example.

uLipe opened the pull request.

microbuilder commented on samples/fusion_imu/src/imu_data_producer.c:
```diff
@@ -0,0 +1,131 @@
+static const struct device *mag_dev = DEVICE_DT_GET(DT_NODELABEL(board_mag));
+static fusion_sensor_data_callback_t imu_user_callback;
+
+/* create the imu measurement header */
+const struct step_mes_header imu_measurement_header = {
+	/* Filter word. */
+	.filter = {
+		.base_type = STEP_MES_TYPE_ORIENTATION,
+		.flags = {
+			.data_format = STEP_MES_FORMAT_NONE,
+			.encoding = STEP_MES_ENCODING_NONE,
+			.compression = STEP_MES_COMPRESSION_NONE,
+			.timestamp = STEP_MES_TIMESTAMP_UPTIME_MS_32,
+		},
+	},
+	/* SI Unit word. */
+	.unit = {
+		.si_unit = STEP_MES_UNIT_SI_METER_PER_SECOND_2,
+		.ctype = STEP_MES_UNIT_CTYPE_IEEE754_FLOAT32,
+		.scale_factor = STEP_MES_SI_SCALE_NONE,
+	},
+	/* Source/Len word. */
+	.srclen = {
+		.len = sizeof(struct imu_data_payload),
+		.fragment = 0,
+		.sourceid = 2,
+	}
+};
```
- Since this is a vector (3 for Euler, 4 for Quaternion), `srclen.vec_sz` should be set to either `3` or `4`.
- `unit.si_unit` is wrong here, but there isn't an obvious existing field to insert today. The closest corresponding SI unit would be `STEP_MES_UNIT_SI_RADIAN`, but that's only useful when we are actually returning values in radians with Euler angles, whereas degrees are more common there.

I've created and merged in a PR (#6) that adds the following values for `.filter.ext_type`:

```c
/** Extended measurement types for STEP_MES_TYPE_ORIENTATION (8-bit). */
enum step_mes_ext_orientation {
	STEP_MES_EXT_TYPE_ORIENTATION_UNDEFINED        = 0, /**< Radian */
	STEP_MES_EXT_TYPE_ORIENTATION_ANG_MEASURE      = 1, /**< Radian or Degree */
	STEP_MES_EXT_TYPE_ORIENTATION_UNIT_QUATERNION  = 2,	/**< AKA versor */
};
```

There is also a new SI unit for degrees:

```c
/** @brief +/- 0.0 ... 359.99 inclusive. Not an offical IS unit for
 *  angular measurements (which should be STEP_MES_UNIT_SI_RADIAN), but
 *  it is listed as an 'accepted' unit. */
STEP_MES_UNIT_SI_DEGREE                 = 0x82,
```

I would propose the following:

| Output Value | `srclen.vec_sz` | `.filter.ext_type` | `.unit.si_unit` |
| - | - | - | - |
| Euler Angles | 3 | STEP_MES_EXT_TYPE_ORIENTATION_ANG_MEASURE | STEP_MES_UNIT_SI_DEGREE |
| Unit Quaternion | 4 | STEP_MES_EXT_TYPE_ORIENTATION_UNIT_QUATERNION | STEP_MES_UNIT_SI_INTERVAL |

microbuilder replied:
Also, the following values can be left out if they are not going to be assigned non-zero values:

- .data_format = STEP_MES_FORMAT_NONE
- .encoding = STEP_MES_ENCODING_NONE
- .compression = STEP_MES_COMPRESSION_NONE
- .scale_factor = STEP_MES_SI_SCALE_NONE
- .fragment = 0

microbuilder replied:
This should end up with something resembling:

```c
/* create the imu measurement header */
const struct step_mes_header imu_measurement_header = {
	/* Filter word. */
	.filter = {
		.base_type = STEP_MES_TYPE_ORIENTATION,
		.ext_type = STEP_MES_EXT_TYPE_ORIENTATION_ANG_MEASURE,
		.flags = {
			.timestamp = STEP_MES_TIMESTAMP_UPTIME_MS_32,
		},
	},
	/* SI Unit word. */
	.unit = {
		.si_unit = STEP_MES_UNIT_SI_RADIAN,
		.ctype = STEP_MES_UNIT_CTYPE_IEEE754_FLOAT32,
		.scale_factor = STEP_MES_SI_SCALE_NONE,
	},
	/* Source/Len word. */
	.srclen = {
		.len = sizeof(struct imu_data_payload),
		.vec_sz = 3,
		.sourceid = 2,
	}
};
```

uLipe replied:
modified according the suggestions.

uLipe commented:
> What do you think is the ideal `.filter.ext_type` and `.unit.si_unit` here?
> 
> Current options are:
> 
> ```c
> /** Extended measurement types for STEP_MES_TYPE_ORIENTATION (8-bit). */
> enum step_mes_ext_orientation {
> 	STEP_MES_EXT_TYPE_ORIENTATION_UNDEFINED        = 0, /**< Radian */
> 	STEP_MES_EXT_TYPE_ORIENTATION_ANG_MEASURE      = 1, /**< Radian or Degree */
> 	STEP_MES_EXT_TYPE_ORIENTATION_UNIT_QUATERNION  = 2,	/**< AKA versor */
> };
> ```
> 
> Default is radians here, but I wonder if it isn't better to pick either unit quaternions (most flexible) or Euler angles in degrees (most understanable to 99% of people)?

I do think the best option would be the native way for processing, not for presenting, since STeP is a data pipeline, I think the best is to keep the most frequent unit for processing, which I think is radians for the most known fusion algorithms.

Since is STeP by design is intented to processing, not presenting, I would keep the radians there, what do you think of my rationale?

microbuilder commented:
> > What do you think is the ideal `.filter.ext_type` and `.unit.si_unit` here?
> > Current options are:
> > ```c
> > /** Extended measurement types for STEP_MES_TYPE_ORIENTATION (8-bit). */
> > enum step_mes_ext_orientation {
> > 	STEP_MES_EXT_TYPE_ORIENTATION_UNDEFINED        = 0, /**< Radian */
> > 	STEP_MES_EXT_TYPE_ORIENTATION_ANG_MEASURE      = 1, /**< Radian or Degree */
> > 	STEP_MES_EXT_TYPE_ORIENTATION_UNIT_QUATERNION  = 2,	/**< AKA versor */
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Default is radians here, but I wonder if it isn't better to pick either unit quaternions (most flexible) or Euler angles in degrees (most understanable to 99% of people)?
> 
> I do think the best option would be the native way for processing, not for presenting, since STeP is a data pipeline, I think the best is to keep the most frequent unit for processing, which I think is radians for the most known fusion algorithms.
> 
> Since is STeP by design is intented to processing, not presenting, I would keep the radians there, what do you think of my rationale?

Sounds fine … I’m actually more curious if quats aren’t more common, though?

uLipe commented:
> > > What do you think is the ideal `.filter.ext_type` and `.unit.si_unit` here?
> > > Current options are:
> > > ```c
> > > /** Extended measurement types for STEP_MES_TYPE_ORIENTATION (8-bit). */
> > > enum step_mes_ext_orientation {
> > > 	STEP_MES_EXT_TYPE_ORIENTATION_UNDEFINED        = 0, /**< Radian */
> > > 	STEP_MES_EXT_TYPE_ORIENTATION_ANG_MEASURE      = 1, /**< Radian or Degree */
> > > 	STEP_MES_EXT_TYPE_ORIENTATION_UNIT_QUATERNION  = 2,	/**< AKA versor */
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Default is radians here, but I wonder if it isn't better to pick either unit quaternions (most flexible) or Euler angles in degrees (most understanable to 99% of people)?
> > 
> > 
> > I do think the best option would be the native way for processing, not for presenting, since STeP is a data pipeline, I think the best is to keep the most frequent unit for processing, which I think is radians for the most known fusion algorithms.
> > Since is STeP by design is intented to processing, not presenting, I would keep the radians there, what do you think of my rationale?
> 
> Sounds fine … I’m actually more curious if quats aren’t more common, though?

AFAIK even proprietary fusion libraries like Bosch BSX uses radians for processing, also the ROS. ROS2 moveit and its navigation stack use radians for the processing, the final result, on the other hand, is delivered in quaternions.

uLipe commented:
Had to remove magnetometer calibration code, since it is producing output matrices with NAN in its elements, even the values of the complimentary filter got better.

Wondering what would the reason, the magnetometer calibratioon (not the fast one) is even worse, since the code hangs inside of it, probably same problem of excessive stack usage leading to overflows.

uLipe commented:
@microbuilder PTAL

microbuilder commented:
The `stream` command is useful, but would we capture CTRL+C or any sort of key input to stop the stream command and recover the shell prompt?

uLipe commented:
Added streaming timed command, since capturing a key would be unecessary extra complicated for this sample context.

Also corrected the PI controller, it seems I pushed wrong values in previous commits.

uLipe commented:
@microbuilder PTAL.
